v1/services/storage/store.go

```diff
@@ -232,26 +233,66 @@ func (s *Store) ReadGroup(ctx context.Context, req *datatypes.ReadGroupRequest)
 	return rs, nil
 }
 
+type metaqueryAttributes struct {
+	orgID      influxdb.ID
+	db, rp     string
+	start, end int64
+	pred       influxql.Expr
+}
+
+func (s *Store) tagKeysWithFieldPredicate(ctx context.Context, mqAttrs *metaqueryAttributes, shardIDs []uint64) (cursors.StringIterator, error) {
+	var cur reads.SeriesCursor
+	if ic, err := newIndexSeriesCursorInfluxQLPred(ctx, mqAttrs.pred, s.TSDBStore.Shards(shardIDs)); err != nil {
+		return nil, err
+	} else if ic == nil {
+		return cursors.EmptyStringIterator, nil
+	} else {
+		cur = ic
+	}
+	m := make(map[string]struct{})
+	rs := reads.NewFilteredResultSet(ctx, mqAttrs.start, mqAttrs.end, cur)
+	for rs.Next() {
+		func() {
+			c := rs.Cursor()
+			if c == nil {
+				// no data for series key + field combination
+				return
+			}
+			defer c.Close()
+			if cursorHasData(c) {
+				for _, tag := range rs.Tags() {
+					m[string(tag.Key)] = struct{}{}
+				}
```

Comment:
Go doesn't optimize these struct loops, which results in copying every `Tag` struct from the slice. I would recommend:

```go
tags := rs.Tags()
for i := range tags {
	m[string(tags[i].Key)] = struct{}{}
}
```